fe/fe-core/src/main/java/com/starrocks/scheduler/TaskManager.java

```diff
@@ -173,14 +240,294 @@ public ShowResultSet handleSubmitTaskStmt(SubmitTaskStmt submitTaskStmt) throws
+        int labelKeepMaxSecond = Config.label_keep_max_second;
+        int numTaskToRemove = currentTask.size() - Config.label_keep_max_num;
+        for (Task task : currentTask) {
+            if ((currentTimeMs - task.getCreateTime()) / 1000 > labelKeepMaxSecond ||
+                    numTaskToRemove > 0) {
+                taskIdToDelete.add(task.getId());
+                --numTaskToRemove;
+            }
+        }
+        // this will do in checkpoint and does not need write log
+        dropTasks(taskIdToDelete, true);
+    }
+
+    public void removeOldTaskRunHistory() {
+        long currentTimeMs = System.currentTimeMillis();
+
+        Deque<TaskRunStatus> taskRunHistory = taskRunManager.taskRunHistory.getAllHistory();
+        List<String> historyToDelete = Lists.newArrayList();
+
+        if (!tryLock()) {
+            return;
+        }
+        try {
+            int labelKeepMaxSecond = Config.label_keep_max_second;
+            int numHistoryToRemove = taskRunHistory.size() - Config.label_keep_max_num;
+            for (TaskRunStatus runStatus : taskRunHistory) {
+                // task run may run for a long time and then failed
+                // there may be no finish time when it fails
+                long lastUpdateTime = runStatus.getCreateTime();
+                if (runStatus.getFinishTime() > lastUpdateTime) {
+                    lastUpdateTime = runStatus.getFinishTime();
+                }
+                if ((currentTimeMs - lastUpdateTime) / 1000 > labelKeepMaxSecond ||
+                        numHistoryToRemove > 0) {
+                    historyToDelete.add(runStatus.getQueryId());
+                    taskRunHistory.remove();
+                    --numHistoryToRemove;
+                }
+            }
+        } finally {
+            unlock();
+        }
+        LOG.info("remove run history:{}", historyToDelete);
+    }
+
+
+    private static class SerializeData {
+        @SerializedName("tasks")
+        public List<Task> tasks;
+
+        @SerializedName("runStatus")
+        public List<TaskRunStatus> runStatus;
+    }
+
+    static class TaskRunManager {
+
+        // taskId -> pending TaskRun Queue, for each Task only support 1 running taskRun currently,
+        // so the map value is FIFO queue
+        private final Map<Long, Queue<TaskRun>> pendingTaskRunMap = Maps.newConcurrentMap();
+
+        // taskId -> running TaskRun, for each Task only support 1 running taskRun currently,
+        // so the map value is not queue
+        private final Map<Long, TaskRun> runningTaskRunMap = Maps.newConcurrentMap();
+
+        // include SUCCESS/FAILED/CANCEL taskRun
+        private final TaskRunHistory taskRunHistory = new TaskRunHistory();
+
+        // Use to execute actual TaskRun
+        private final TaskRunExecutor taskRunExecutor = new TaskRunExecutor();
+
+        public SubmitResult submitTaskRun(TaskRun taskRun) {
```

Comment:
This class should be protected by a lock?
why not use a Future to check if the task finish.

Reply:
SubmitResult isSubmitResult is an asynchronous state used to indicate whether the submission is successful. The future can only check isCancelled or isDone. When the future task is running, the state of SubmitResult is actually Submitted.I think this state does not needs to be locked.